### PR TITLE
[TRUST] Ferreous Coffin

### LINKED ISSUE #431: There is no Gambit targetting option for dead party members
https://github.com/LandSandBoat/server/blob/base/src/map/ai/helpers/gambits_container.cpp#L261-L273

A new section like this would do the job:
```cpp
                // New G_TARGET enum
                else if (gambit.predicates[0].target == G_TARGET::PARTY_DEAD)
                {
                    // Is in combat
                    if (auto* PMob = dynamic_cast<CMobEntity*>(POwner->GetBattleTarget()))
                    {
                        static_cast<CCharEntity*>(POwner->PMaster)->ForParty([&](CBattleEntity* PMember) {
                            if (/* Do regular checking, but invert the logic to select dead party members */)
                            {
                                target = PMember;
                            }
                        });
                    }
                }
```

- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have searched existing [issues](https://github.com/LandSandBoat/server/issues) to see if the issue has already been opened
- [x] I have checked the commit log to see if the issue has been resolved since my server was last updated
- [x] I have read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I have specified what branch this happens on branch: EDITME

**_Additional Information_** (Steps to reproduce/Expected behavior) **:**

### PR DESCRIPTION
**Pull request by neuromancerxi**
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

Updates Trust Script for Ferreous Coffin
Updates Gambits, Mob Spell Container, and Spell CPP.
Updates mob skills, spells SQL.

Closes #431

## Steps to test these changes

Build a Party of at least two
Cast Ferreous Coffin
Engage a mob
Watch standard WHM behavior
`!hp 0` your 2nd party member
Observe Ferreous Coffin cast the highest available Raise on your fallen party member.

### CODE REVIEW

**Comment by neuromancerxi** [comment]
~~Currently as a WIP due to being unable to get Ferreous Coffin to Cast any Raise spell on a party member that is KO'd.~~

~~I've tried `HPP_LT` and `STATUS` `KO` conditions, and the current implementation of the gambit `IS_DEAD` doesn't seem to work either. I'm thinking there's an `->isDead` check happening somewhere before it gets to the gambit stack, but not sure where to look. Would appreciate a 2nd set of eyes on this if possible.~~

~~All other items work as advertised on the Tin.~~

Now working as expected thanks to the details listed in #431 and Clay pointing me to it via linking the Issue. Thanks Zach and Clay.

**Comment by neuromancerxi** [comment]
Built Target Option for Dead Party Members thanks to the linked issue, added commit with Zach as co-author since it's primarily his code. Gambit condition for Dead Party Member now works appropriately.

File: scripts/globals/spells/trust/ferreous_coffin.lua [hunk 1]
```diff
@@ -1,5 +1,12 @@
 -----------------------------------
 -- Trust: Ferreous Coffin
+-- Auto Refresh II (Cleric's Bliaut +2). HP-10%, MP+35%
```
Comment (WinterSolstice8) [inline scripts/globals/spells/trust/ferreous_coffin.lua:3]: Interesting implication of visible gear effecting trusts' stats. I wonder if there are more? Some gear like AF should be more easily verifiable, just worth a thought when implementing trusts. Too bad Ferreous Coffin doesn't cast regen, we could verify Regen potency

File: scripts/globals/spells/trust/ferreous_coffin.lua [hunk 2]
```diff
@@ -15,6 +22,46 @@ end
...
+
+    mob:addSimpleGambit(ai.t.PARTY_DEAD, ai.c.ALWAYS, 0, ai.r.MA, ai.s.HIGHEST, xi.magic.spellFamily.RAISE)
+
+    mob:addListener("WEAPONSKILL_USE", "FERREOUS_COFFIN_WEAPONSKILL_USE", function(mobArg, target, wsid, tp, action)
+        if wsid == 170 then -- Randgrith
+        -- Return to the dust whence you came! Randgrith!!!
+            if math.random(100) <= 66 then
+                xi.trust.message(mobArg, xi.trust.message_offset.SPECIAL_MOVE_1)
+            end
+            mob:addStatusEffect(xi.effect.ACCURACY_BOOST, 20, 0, 20) -- Cheat in Relic AM ACC
```
Comment (WinterSolstice8) [inline scripts/globals/spells/trust/ferreous_coffin.lua:58]: Do we get an accurate `tp` value in here? could get the duration with `math.floor(tp*0.02)` for duration as if it were a real relic

Comment (neuromancerxi) [reply scripts/globals/spells/trust/ferreous_coffin.lua:58]: Considering his WS Rule (`ASAP`), he'll be swinging at ~1k TP each time, so I targeted Relic AM duration and ACC values based on that.

**Comment by WinterSolstice8** [comment]
I'm curious to see if Ferreous Coffin's Mjollnir does the MP absorb add effect, going to test it in retail

**Comment by neuromancerxi** [comment]
> I'm curious to see if Ferreous Coffin's Mjollnir does the MP absorb add effect, going to test it in retail

That would be slick if so. Do we have a mechanism to reliably apply something like that to a Trust script? I ran into a wall trying to apply Algol effects to Rughadjeen's script.

**Comment by TeoTwawki** [comment]
> I'm curious to see if Ferreous Coffin's Mjollnir does the MP absorb add effect, going to test it in retail

it doesn't take it from anywhere, just grants mp. thats why it works vs undead. I dunno if mr. Coffin here procs tho, or if he does if its same rate as a players. I just know its recovery not really absorb. :)

**Comment by WinterSolstice8** [comment]
I didn't see any 3x damage or MP additional effects, too bad. That would've been a decent bonus for a trust like this.

**Comment by WinterSolstice8** [comment]
>  I ran into a wall trying to apply Algol effects to Rughadjeen's script.

~~Offhand, this is what I would try. Untested, of course. The "weapon" should be populated so:~~

```
local weapon = mob:getEquippedItem(xi.slot.MAIN)
if weapon then
    weapon:addMod(xi.mod.ITEM_ADDEFFECT_TYPE, 1)
    weapon:addMod(xi.mod.ITEM_SUBEFFECT, 1)
    weapon:addMod(xi.mod.ITEM_ADDEFFECT_DMG, 33)
    weapon:addMod(xi.mod.ITEM_ADDEFFECT_CHANCE, 33)
    weapon:addMod(xi.mod.ITEM_ADDEFFECT_ELEMENT, 1)
end
```

~~It looks like you'll have to edit battleutils.cpp:~~
https://github.com/LandSandBoat/server/blob/41519a4d94e41e8a5b6770abaf3f959e44bf8b27/src/map/utils/battleutils.cpp#L1320
~~hardcoded PC check prevents the addeffect calls for trusts~~

EDIT: wrong lua call to get equipped item
edit2: did not know this was a bad idea

**Review by WinterSolstice8 — APPROVED** [review]

**Comment by TeoTwawki** [comment]
> stuff winter suggested

nope. Trusts you'd set up the way mobs are, because technically they kind of are. [Take a look in an existing monster that has additional effects.](https://github.com/LandSandBoat/server/blob/2dbae3a1d7e3de544670f35051620660d5f9205f/scripts/zones/Attohwa_Chasm/mobs/Sargas.lua) Mobs (and trusts) aren't supposed to have actual weapons, and even if you try to cast their dummy "weapon" to one it won't have the guts to store mods.

**Comment by TeoTwawki** [comment]
> I didn't see any 3x damage or MP additional effects, too bad. That would've been a decent bonus for a trust like this.

`#sadTeo`

**Comment by WinterSolstice8** [comment]
~~P.S.
getEquippedItem also has a PC hardcoded check, which makes sense here.~~
https://github.com/LandSandBoat/server/blob/41519a4d94e41e8a5b6770abaf3f959e44bf8b27/src/map/lua/lua_baseentity.cpp#L3265

~~Suggestion: add "getTrustEquippedWeapon", make it only work for trust and main/ammo/ranged slots. This way we could add additional effects like Holy Bolts to a theoretical module trust, or other add effects. It's that or remove the PC check, but I think that doesn't make much sense. Trusts (currently) don't need armors.~~

edit: bad idea

**Comment by TeoTwawki** [comment]
> P.S. getEquippedItem also has a PC hardcoded check, which makes sense here.
> 
> https://github.com/LandSandBoat/server/blob/41519a4d94e41e8a5b6770abaf3f959e44bf8b27/src/map/lua/lua_baseentity.cpp#L3265
> 
> Suggestion: add "getTrustEquippedWeapon", make it only work for trust and main/ammo/ranged slots. This way we could add additional effects like Holy Bolts to a theoretical module trust, or other add effects. It's that or remove the PC check, but I think that doesn't make much sense. Trusts (currently) don't need armors.

no really, don't do this

**Comment by WinterSolstice8** [comment]
Wrote up that reply before I saw yours Teo, yeah, don't do that

**Review by claywar — APPROVED** [review]

**Pull request merged**